File: java-spring/src/main/java/org/burgas/javaspring/mapper/IdentityMapper.java
package org.burgas.javaspring.mapper;

import lombok.RequiredArgsConstructor;
import org.burgas.javaspring.dto.identity.IdentityFullResponse;
import org.burgas.javaspring.dto.identity.IdentityRequest;
import org.burgas.javaspring.dto.identity.IdentityShortResponse;
import org.burgas.javaspring.entity.identity.Authority;
import org.burgas.javaspring.entity.identity.Identity;
import org.burgas.javaspring.repository.IdentityRepository;
import org.springframework.beans.factory.ObjectFactory;
import org.springframework.stereotype.Component;

import java.util.ArrayList;
import java.util.Optional;
import java.util.UUID;


@Component
@RequiredArgsConstructor
public final class IdentityMapper implements Mapper<IdentityRequest, Identity, IdentityShortResponse, IdentityFullResponse> {

    public final IdentityRepository identityRepository;
    private final ObjectFactory<NoteMapper> noteMapperObjectFactory;

    private NoteMapper getNoteMapper() {
        return this.noteMapperObjectFactory.getObject();
    }

    @Override
    public Identity toEntity(IdentityRequest request) {
        return this.identityRepository.findById(request.getId() == null ? new UUID(0, 0) : request.getId())
                .map(
                        identity -> Identity.builder()
                                .id(identity.getId())
                                .authority(Optional.ofNullable(request.getAuthority()).orElse(identity.getAuthority()))
                                .username(Optional.ofNullable(request.getUsername()).orElse(identity.getUsername()))
                                .password(identity.getPassword())
                                .email(Optional.ofNullable(request.getEmail()).orElse(identity.getEmail()))
                                .enabled(identity.getEnabled())
                                .firstname(Optional.ofNullable(request.getFirstname()).orElse(identity.getFirstname()))
                                .lastname(Optional.ofNullable(request.getLastname()).orElse(identity.getLastname()))
                                .patronymic(Optional.ofNullable(request.getPatronymic()).orElse(identity.getPatronymic()))
                                .build()
                )
                .orElseGet(
                        () -> Identity.builder()
                                .authority(Optional.ofNullable(request.getAuthority()).orElse(Authority.USER))
                                .username(Optional.ofNullable(request.getUsername()).orElseThrow())
                                .password(Optional.ofNullable(request.getPassword()).orElseThrow())
                                .email(Optional.ofNullable(request.getEmail()).orElseThrow())
                                .enabled(Optional.ofNullable(request.getEnabled()).orElse(true))
                                .firstname(Optional.ofNullable(request.getFirstname()).orElseThrow())
                                .lastname(Optional.ofNullable(request.getLastname()).orElseThrow())
                                .patronymic(Optional.ofNullable(request.getPatronymic()).orElseThrow())
                                .build()
                );
    }

    @Override
    public IdentityShortResponse toShortResponse(Identity entity) {
        return IdentityShortResponse.builder()
                .id(entity.getId())
                .username(entity.getUsername())
                .email(entity.getEmail())
                .firstname(entity.getFirstname())
                .lastname(entity.getLastname())
                .patronymic(entity.getPatronymic())
                .image(entity.getImage())
                .build();
    }

    @Override
    public IdentityFullResponse toFullResponse(Identity entity) {
        return IdentityFullResponse.builder()
                .id(entity.getId())
                .username(entity.getUsername())
                .email(entity.getEmail())
                .firstname(entity.getFirstname())
                .lastname(entity.getLastname())
                .patronymic(entity.getPatronymic())
                .image(entity.getImage())
                .notes(
                        Optional.ofNullable(entity.getNotes())
                                .map(notes -> notes.stream()
                                        .map(note -> this.getNoteMapper().toShortResponse(note)).toList())
                                .orElseGet(ArrayList::new)
                )
                .build();
    }
}

File: kotlin-spring/src/main/kotlin/org/burgas/kotlinspring/mapper/IdentityMapper.kt
package org.burgas.kotlinspring.mapper

import org.burgas.kotlinspring.entity.identity.Authority
import org.burgas.kotlinspring.entity.identity.Identity
import org.burgas.kotlinspring.entity.identity.IdentityFullResponse
import org.burgas.kotlinspring.entity.identity.IdentityRequest
import org.burgas.kotlinspring.entity.identity.IdentityShortResponse
import org.burgas.kotlinspring.repository.IdentityRepository
import org.springframework.beans.factory.ObjectFactory
import org.springframework.security.crypto.password.PasswordEncoder
import org.springframework.stereotype.Component
import java.util.UUID

@Component
class IdentityMapper : Mapper<IdentityRequest, Identity, IdentityShortResponse, IdentityFullResponse> {

    final val identityRepository: IdentityRepository
    private final val noteMapperObjectFactory: ObjectFactory<NoteMapper>
    private final val passwordEncoder: PasswordEncoder

    constructor(
        identityRepository: IdentityRepository,
        noteMapperObjectFactory: ObjectFactory<NoteMapper>,
        passwordEncoder: PasswordEncoder
    ) {
        this.identityRepository = identityRepository
        this.noteMapperObjectFactory = noteMapperObjectFactory
        this.passwordEncoder = passwordEncoder
    }

    private fun getNoteMapper(): NoteMapper = this.noteMapperObjectFactory.`object`

    override fun toEntity(request: IdentityRequest): Identity {
        return this.identityRepository.findById(request.id ?: UUID(0, 0))
            .map {
                Identity().apply {
                    this.id = it.id
                    this.authority = request.authority ?: it.authority
                    this.username = request.username ?: it.username
                    this.password = it.password
                    this.email = request.email ?: it.email
                    this.enabled = request.enabled ?: it.enabled
                    this.firstname = request.firstname ?: it.firstname
                    this.lastname = request.lastname ?: it.lastname
                    this.patronymic = request.patronymic ?: it.patronymic
                    this.image = it.image
                }
            }
            .orElseGet {
                Identity().apply {
                    val newPassword =
                        if (!request.password.isNullOrEmpty()) passwordEncoder.encode(request.password) else throw IllegalArgumentException(
                            "Identity password is null or empty"
                        )
                    this.authority = request.authority ?: Authority.USER
                    this.username = request.username ?: throw IllegalArgumentException("Identity username is null")
                    this.password = newPassword ?: throw IllegalArgumentException("Password not encoded")
                    this.email = request.email ?: throw IllegalArgumentException("Identity email is null")
                    this.enabled = request.enabled ?: true
                    this.firstname = request.firstname ?: throw IllegalArgumentException("Identity firstname is null")
                    this.lastname = request.lastname ?: throw IllegalArgumentException("Identity lastname is null")
                    this.patronymic =
                        request.patronymic ?: throw IllegalArgumentException("Identity patronymic is null")
                }
            }
    }

    override fun toShortResponse(entity: Identity): IdentityShortResponse {
        return IdentityShortResponse(
            id = entity.id,
            username = entity.username,
            email = entity.email,
            firstname = entity.firstname,
            lastname = entity.lastname,
            patronymic = entity.patronymic,
            image = entity.image
        )
    }

    override fun toFullResponse(entity: Identity): IdentityFullResponse {
        return IdentityFullResponse(
            id = entity.id,
            username = entity.username,
            email = entity.email,
            firstname = entity.firstname,
            lastname = entity.lastname,
            patronymic = entity.patronymic,
            image = entity.image,
            notes = entity.notes.map { this.getNoteMapper().toShortResponse(it) }
        )
    }
}